TensorRT/plugins/HcalProducerFPGA.cc

```diff
@@ -0,0 +1,163 @@
+			      current=strtok(cstr," ");
+			      while(current!=NULL){
+				fInVals.push_back(atof(current));
+				current=strtok(NULL," ");
+			      }
+			      cstr=const_cast<char*>(pline.c_str());
+			      current=strtok(cstr," ");
+			      while(current!=NULL){
+				fPrVals.push_back(atof(current));
+				current=strtok(NULL," ");
+			      }
+			    }
+			  }
+	                }			
+			fCount = 0;
+			*/
+			
+		}
+		void acquire(edm::Event const& iEvent, edm::EventSetup const& iSetup, Input& iInput) override {
+			auto ninput = client_.ninput();
+			auto batchSize = client_.batchSize();
+			iInput = Input(ninput*batchSize, 0.f);
+			//make some random channels
+			for(unsigned ib = 0; ib < batchSize; ib++) { 
+			  //Current Hcal setup takes 11 inputs but it is sent in chunks of 16 to the FPGA 
+			  iInput[ib*ninput+0] = f_to_ui<16,6>(1);
+			  iInput[ib*ninput+1] = f_to_ui<16,6>(2);
+			  iInput[ib*ninput+3] = f_to_ui<16,6>(int(rand() % 30)-15); 
+			  iInput[ib*ninput+4] = f_to_ui<16,6>(int(rand() % 36)-36); 
+			  for(unsigned i0 = 5; i0 < 18; i0++) iInput[ib*ninput+i0] = f_to_ui<16,8>(1);
+			  for(unsigned i0 = 18; i0 < ninput; i0++) iInput[ib*ninput+i0] = f_to_ui<16,8>(0);
+			  /* for debug
+			  for(unsigned i0 = 0;  i0 < 11; i0++)     iInput[ib*ninput+i0] = f_to_ui<32,14>(fInVals[i0+11*fCount]);
+			  for(unsigned i0 = 11; i0 < ninput; i0++) iInput[ib*ninput+i0] = 0;
+			  fCount++;
+			  */
+			}
+		}
+		void produce(edm::Event& iEvent, edm::EventSetup const& iSetup, Output const& iOutput) override {
+			//check the results
+			findTopN(iOutput);
+		}
+		~HcalProducerFPGA() override {}
+		//to ensure distinct cfi names - specialized below
+		static std::string getCfiName();
+		static void fillDescriptions(edm::ConfigurationDescriptions & descriptions) {
+			edm::ParameterSetDescription desc;
+			Client::fillPSetDescription(desc);
+			desc.add<unsigned>("topN",5);
+			descriptions.add(getCfiName(),desc);
+		}
+		//Just putting something in for the hell of it
+		void findTopN(const Output& scores) const {
+			auto dim = client_.noutput();
+			//int batchsize = client_.batchSize();
+			for(unsigned i0 = 0; i0 < client_.batchSize(); i0++) {
+				//match score to type by index, then put in largest-first map
+				std::map<float,std::string,std::greater<float>> score_map;
+				for(unsigned i = 0; i < (unsigned)dim; ++i){
+					std::stringstream pSS; pSS << "Dummy Channel " << i;
+					score_map.emplace(scores[i0*dim+i],pSS.str());
+				}
+				//get top n
+				std::stringstream msg;
+				msg << "Scores:\n";
+				unsigned counter = 0;
+				for(const auto& item: score_map){
+					msg << item.second << " : " << item.first << "\n";
+					++counter;
+					if(counter>=topN_) break;
+				}
+				edm::LogInfo("HcalProducerFPGA") << msg.str();
+			}
+		}
+
+
+	private:
+		using SonicEDProducer<Client>::client_;
+                template<unsigned int B, unsigned int I>
+                unsigned short f_to_ui(float f) {
```

Comment:
My proposal is to make this a function of the client rather than the producer, and add a parameter to the client to turn on this conversion. Then the existing TRTClient can be used, with just a modification to use the "derived" input variable.

Alternatively, it seems like the TRTClientFPGA could be defined directly from the existing client as:
```
typedef TRTClient<SonicClientSync<std::vector<unsigned short>>> TRTClientFPGASync;
```
etc.